src/execution.jl

```diff
@@ -52,3 +52,89 @@ function assign_args!(code, args)
 
     return vars, var_exprs
 end
+
+
+## cached compilation
+
+const cache_lock = ReentrantLock()
+
+"""
+    cached_compilation(cache::Dict{UInt}, cfg::CompilerConfig, ft::Type, tt::Type,
+                       compiler, linker)
+
+Compile a method instance, identified by its function type `ft` and argument types `tt`,
+using `compiler` and `linker`, and store the result in `cache`.
+
+The `cache` argument should be a dictionary that can be indexed using a `UInt` and store
+whatever the `linker` function returns. The `compiler` function should take a `CompilerJob`
+and return data that can be cached across sessions (e.g., LLVM IR). This data is then
+forwarded, along with the `CompilerJob`, to the `linker` function which is allowed to create
+session-dependent objects (e.g., a `CuModule`).
+"""
+function cached_compilation(cache::AbstractDict{UInt,V},
+                            cfg::CompilerConfig,
+                            ft::Type, tt::Type,
+                            compiler::Function, linker::Function) where {V}
+    # NOTE: we index the cach both using (ft, tt, world, cfg) keys, for the fast look-up,
+    #       and using CodeInfo keys for the slow look-up. we need to cache both for
+    #       performance, but cannot use a separate private cache for the ci->obj lookup
+    #       (e.g. putting it next to the CodeInfo's in the CodeCache) because some clients
+    #       expect to be able to wipe the cache (e.g. CUDA.jl's device_reset!)
+
+    # fast path: index the cache directly for the *current* world + kernel + compiler config
+
+    world = tls_world_age()
+    key = hash(ft)
+    key = hash(tt, key)
```

Comment:
These may often return `0`, to indicate that the hash function is uncomputable. You should key the cache on the objects themselves, not an unreliable function of them. A MethodInstance object is likely ideal for that, since it layers a stable hash identity on top of the combination of types and Methods, further resolving the issue that `tt` does not uniquely represent a Method, but rather represents a set of Method of unpredictable length.

Reply:
I did it this way because looking up the `mi` isn't cheap, so I'd rather not do that on every kernel launch.

Reply:
The mi are held in a hash table

Reply:
So `jl_gf_invoke_lookup` should be fast, or which entrypoint should we be using? I guess I might be doing something wrong in https://github.com/JuliaGPU/GPUCompiler.jl/blob/5587ba154472d6a0d76b5fc242d45ca9fcbdd54f/src/jlgen.jl#L223-L247 too, that code has grown organically.

Reply:
Yes, that is good too. The internals use MethodInstance usually in the cache for that reason too. MethodMatch itself probably is not quite desirable as it might (in very rare cases) violate the egal output requirements

Reply:
OK, works great, thanks for the suggestion. I did still end up storing `objectid(id)` in the cache, because otherwise we allocate too much in the hot path (as we're storing a tuple of `(mi, world, compiler_cfg)`).